File: packages/actor-core/src/manager/protocol/query.ts
import { ActorKeySchema, type ActorKey } from "@/common//utils";
import { z } from "zod";
import { EncodingSchema } from "@/actor/protocol/serde";

export const CreateRequestSchema = z.object({
	name: z.string(),
	key: ActorKeySchema,
	region: z.string().optional(),
});

export const GetForKeyRequestSchema = z.object({
	name: z.string(),
	key: ActorKeySchema,
});

export const GetOrCreateRequestSchema = z.object({
	name: z.string(),
	key: ActorKeySchema,
	region: z.string().optional(),
});

export const ActorQuerySchema = z.union([
	z.object({
		getForId: z.object({
			actorId: z.string(),
		}),
	}),
	z.object({
		getForKey: GetForKeyRequestSchema,
	}),
	z.object({
		getOrCreateForKey: GetOrCreateRequestSchema,
	}),
	z.object({
		create: CreateRequestSchema,
	}),
]);

export const ConnectQuerySchema = z.object({
	query: ActorQuerySchema,
	encoding: EncodingSchema,
	params: z.string().optional(),
});

export type ActorQuery = z.infer<typeof ActorQuerySchema>;
export type GetForKeyRequest = z.infer<typeof GetForKeyRequestSchema>;
export type GetOrCreateRequest = z.infer<typeof GetOrCreateRequestSchema>;
export type ConnectQuery = z.infer<typeof ConnectQuerySchema>;
/**
 * Interface representing a request to create an actor.
 */
export type CreateRequest = z.infer<typeof CreateRequestSchema>;

File: clients/rust/src/protocol/query.rs
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

use crate::common::ActorKey;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateRequest {
    pub name: String,
    pub key: ActorKey,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<JsonValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetForKeyRequest {
    pub name: String,
    pub key: ActorKey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetForIdRequest {
    #[serde(rename = "actorId")]
    pub actor_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GetOrCreateRequest {
    pub name: String,
    pub key: ActorKey,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<JsonValue>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ActorQuery {
    GetForId {
        #[serde(rename = "getForId")]
        get_for_id: GetForIdRequest,
    },
    GetForKey {
        #[serde(rename = "getForKey")]
        get_for_key: GetForKeyRequest,
    },
    GetOrCreateForKey {
        #[serde(rename = "getOrCreateForKey")]
        get_or_create_for_key: GetOrCreateRequest,
    },
    Create {
        create: CreateRequest,
    },
}